### DISCUSSION TITLE
onStart callback?

### DISCUSSION THREAD

**Discussion by drscottlobo** [danigb/smplr#18] [Ideas]
Sorry! Should have requested both at once, could you also add a callback when scheduled audio starts? onStart: callback?

Thanks, this has been the nicest MIDI library I've found yet!

**Comment by danigb** [comment]
Hi, this is more tricky to implement, and I'm not sure if I want to. In theory, you already know when things start (not always the same with the stop time). May I ask you why you need it?

**Comment by drscottlobo** [comment]
Hi! So I’m doing some basic score notation in React where a series of notes is played back for the user, and like Sibelius or MuseScore, as the notes playback I want each note in the score to change color.  

So what I imagine to be the easiest solution is to have a setHighlight called when each note starts and a setHighlight when each ends. I have tried a setTimeout in the past but it’s totally unreliable and the highlights end up out of sync. I’ve been using waaclock to do it better. But thought if smplr was taking care of the timing, it might be easy to just fire those events onStart and (already working great) onEnded. Maybe set timeout works more reliably somehow when using the AudioContext? 

Thanks for your help and any suggestions.

**Reply by danigb** [reply]
Hi, thanks for the explanation. I'm still unsure if I should add this to smplr, just because I feel is not part of what a sampler should do (and schedule-related things is something quite complex and requirements escalates fast). 

Meanwhile, you can use something like this:
```js
const ctx = new AudioContext();

function triggerAtTime(time, callback) {
  const node = new ConstantSourceNode(ctx);
  node.onended = callback;
  node.start();
  node.stop(time);
}

triggerAtTime(ctx.currentTime + 0.1, () => console.log("Hi"));
```
But I don't know if that approach works well if the number of notes is very large (one of the potential problems of adding this kind of code to smplr)

**Reply by drscottlobo** [reply]
Hi @danigb! This works just fine! I'm mainly doing very short musical examples (never more than say 50 notes max but mostly just a scale or a couple 4-part chords for basic four-voice part-writing), so far this seems to work just fine. It's so nice to see the ScriptProcessorNode deprecation warning go away!
